### PR TITLE
Combination

### PR DESCRIPTION
**Pull request by Farnazmdi**
2 functions added.
1. `combination`, that finds the number of cells in G1 and G2, given a set of ODE parameters for a specific drug concentration. It creates a paramRange (like the one we had in sensitivity). Returns the # cells in g1, # cells in g2, and the parameter range.
2. `central_difference`. I wrote the central difference function to get the derivative of the number of cells in each phase with respect to the parameter range.

I guess we need to define a threshold. I plotted them almost all, for alpha, beta, gamma1, and gamma2. The behavior mostly makes sense except for beta that has 1 or 2 small jumps at the beginning.

### CODE REVIEW

**Comment by aarmey** [comment]
Can you post what you plotted? I'm not following what you did here.

You definitely don't want to implement central differencing yourself. Use `Calculus.finite_difference()` or `FiniteDiff.finite_difference_gradient()`

**Comment by Farnazmdi** [comment]
This is the # of cells in G1 vs. a range of gamma2.
image

And this is the derivative of above w.r.t. gamma2:
image

The problem with using `Calculus` was it needed a function, but I just had data points. I tried to make a function but it didn't work, so I just wrote what it was on the Wikipedia. I notice it needs a little bit of change though fo the x-axis.
Is this wrong?

**Comment by aarmey** [comment]
But you just need a function that takes in a set of parameters, and outputs the number of cells at `t=T`. Isn't `predict()` 99% of that?

**Comment by aarmey** [comment]
I think you're differencing the same parameter with different intervals... not sure how to use that. For example, what is the x-axis on your plot? What does that mean biologically?

**Comment by Farnazmdi** [comment]
Yes sorry, you are right. This is useless. I thought the x-axis indirectly will tell us about the concentration. 

What we want is to get the partial derivative of a function -- that is a function of all parameters and `T` -- with respect to one of the parameters.
Is it correct?

**Comment by aarmey** [comment]
Yep.

File: src/Hill.jl [hunk 1]
```diff
@@ -134,3 +134,32 @@ function plotUnitSensitivity(paramRange, result, realParam, i)
     plot!([realParam], seriestype = "vline", margin = 0.3cm, legend = :false)
     ylims!((1E2, 1E4))
 end
+
+""" Calculate the # of cells in G1 for a set of parameters and T """
+function numG1(params, g0, T)
```
Comment (aarmey) [inline src/Hill.jl]: Except you don't want numG1 or numG2, you want number of total cells. You should just have one function with these added together.

**Comment by Farnazmdi** [comment]
Should I calculate this for all the combinations of the parameters?

**Comment by aarmey** [comment]
We should look at a plot with drug concentration on the x-axis, and d[#]/d[alpha], d[#]/d[beta],  d[#]/d[gamma1], and d[#]/d[gamma2] as the y-axis quantities.

**Comment by Farnazmdi** [comment]
doxorubicin:
image

lapatinib:
image

gemcitabine:
image

paclitaxel:
image

**Comment by aarmey** [comment]
First, concentration should be a log axis. Second, looks like alpha/beta should probably be on a separate plot. Third, it might work better to plot d[number]/d[param]/number. I.e. normalize the derivative to the number of cells, so you’re looking at the proportional effect.

**Comment by Farnazmdi** [comment]
Sorry I bothered you a lot in this PR.
I couldn't figure out how to interpret this. 
I understand that we want the combination of the two drugs that one of them decreases gamma1 (for example) and the other decreases gamma2. By decrease, I mean more negative in this plot. The more negative the gradient is, the more the cells are dying? 
I guess `beta` could be seen as the division rate, but do we want the same thing as above with the transition rate, too? I mean do we want to pick those two drugs that one decreases `alpha` and the other decreases `beta`?

Sorry I can't get a sense of the derivative of a function -- with some parameters -- that returns the number of something, with respect to one of its parameters.
I have this analogy: 
The velocity is a function of time, and when we write it as a linear function, acceleration is one of the parameters in the formula.
The number of cells is a function of concentration, but we have transition and death rates as its parameters. 
I don't get what is meant by derivative of velocity wrt the acceleration.


doxorubicin:
image

gemcitabine:
image

lapatinib:
image

paclitaxel:
image

**Comment by aarmey** [comment]
I don't follow your velocity vs acceleration analogy. Velocity is the first derivative, acceleration is the second derivative. We have no second derivatives here.

These are the direction of change effect on cell # when increasing the indicated parameter.

**Comment by Farnazmdi** [comment]
Yes sorry, it is not a good example. I meant v(t) = at + v0 ==> dv(t)/da.

So it means for each concentration (or a set of parameters), when we increase a parameter by one unit, how this affects the cell #. Is that what it is?

**Comment by aarmey** [comment]
I think that's a reasonable description. So, for example, it looks like you should combine lapatinib with a drug that induces gamma1, not gamma2. Or, a drug that inhibits alpha over one that inhibits beta.

**Comment by Farnazmdi** [comment]
Thank you.

File: src/Hill.jl [hunk 2]
```diff
@@ -134,3 +134,38 @@ function plotUnitSensitivity(paramRange, result, realParam, i)
     plot!([realParam], seriestype = "vline", margin = 0.3cm, legend = :false)
     ylims!((1E2, 1E4))
 end
+
+
+""" Calculate the # of cells in G1 for a set of parameters and T """
+function numcells(params, g0, T)
+    t = LinRange(0.0, 95.5, 3)
+    temp = copy(params)
```
Comment (aarmey) [inline src/Hill.jl]: Is there a reason you need to copy params?

File: src/Hill.jl [hunk 3]
```diff
@@ -134,3 +134,38 @@ function plotUnitSensitivity(paramRange, result, realParam, i)
...
+    diffcells(x) = numcells(x, g0, T)
+    return Calculus.finite_difference(diffcells, params)
+end
+
+
+""" Plot the gradient vs concentrations """
+function plotGradient(effects, concentration, g0, T)
+    dif = zeros(4,8)
+    for i = 1:8
+        params = effects[:, i]
```
Comment (aarmey) [inline src/Hill.jl]: Why copy out effects into the `param` variable?

File: src/Hill.jl [hunk 4]
```diff
@@ -134,3 +134,38 @@ function plotUnitSensitivity(paramRange, result, realParam, i)
...
+        num = numcells(params, g0, T)
```
Comment (aarmey) [inline src/Hill.jl]: Why not do `numcells` correction inside `diffCell()`?

File: src/Hill.jl (hunk 2 above)
Comment (Farnazmdi) [reply src/Hill.jl]: Oh this has remained from before that I wanted to sweep the parameters. I will remove it.

File: src/Hill.jl (hunk 3 above)
Comment (Farnazmdi) [reply src/Hill.jl]: I will correct it.

File: src/Hill.jl (hunk 4 above)
Comment (Farnazmdi) [reply src/Hill.jl]: Will do right away.

Comment (Farnazmdi) [reply src/Hill.jl]: Wait, sorry what correction? I guess I need the function. You mean I use `predict` as the function to pass to `Calculus...`?

Comment (aarmey) [reply src/Hill.jl]: I can do this after you do the others. Just ping me.

Comment (Farnazmdi) [reply src/Hill.jl]: Sorry, ignore it.

Comment (Farnazmdi) [reply src/Hill.jl]: I made the edits. I took the normalization into the `diffCell()` but I'm not sure this addresses your point.

**Pull request merged**